Security: audit event not properly logged/audited

## Environment Details
* Helidon Version: 1.1.2
* Helidon MP
* JDK version: 11.0.2
* OS: Windows 10
* Docker version (if applicable):

----------

## Problem Description
Audit event can fail, when number of expected parameters does not match the number of real parameters.

`AM io.helidon.microprofile.server.ServerImpl$1 toResponse
WARNING: Internal server error
io.helidon.security.SecurityException: Failure while executing asynchronous security
	at io.helidon.security.SecurityResponse.get(SecurityResponse.java:67)
	at io.helidon.security.SecurityClient.get(SecurityClient.java:46)
	at io.helidon.security.SecurityClientBuilder.buildAndGet(SecurityClientBuilder.java:61)
	at io.helidon.security.integration.jersey.SecurityFilterCommon.processAuthorization(SecurityFilterCommon.java:313)
	at io.helidon.security.integration.jersey.SecurityFilterCommon.authorize(SecurityFilterCommon.java:299)
	at io.helidon.security.integration.jersey.SecurityFilter.processSecurity(SecurityFilter.java:184)
	at io.helidon.security.integration.jersey.SecurityFilterCommon.doFilter(SecurityFilterCommon.java:140)
	at io.helidon.security.integration.jersey.SecurityFilter.filter(SecurityFilter.java:155)
	at org.glassfish.jersey.server.ContainerFilteringStage.apply(ContainerFilteringStage.java:108)
	at org.glassfish.jersey.server.ContainerFilteringStage.apply(ContainerFilteringStage.java:44)
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:173)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:245)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:679)
	at io.helidon.webserver.jersey.JerseySupport$JerseyHandler.lambda$doAccept$3(JerseySupport.java:265)
	at io.helidon.common.context.Contexts.runInContext(Contexts.java:98)
	at io.helidon.common.context.ContextAwareExecutorImpl.lambda$wrap$1(ContextAwareExecutorImpl.java:129)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.util.concurrent.ExecutionException: java.util.MissingFormatArgumentException: Format specifier '%s'
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2022)
	at io.helidon.security.SecurityResponse.get(SecurityResponse.java:63)
	... 25 more
Caused by: java.util.MissingFormatArgumentException: Format specifier '%s'
	at java.base/java.util.Formatter.format(Formatter.java:2672)
	at java.base/java.util.Formatter.format(Formatter.java:2609)
	at java.base/java.lang.String.format(String.java:2897)
	at io.helidon.security.DefaultAuditProvider.formatMessage(DefaultAuditProvider.java:109)
	at io.helidon.security.DefaultAuditProvider.logEvent(DefaultAuditProvider.java:96)
	at io.helidon.security.DefaultAuditProvider.audit(DefaultAuditProvider.java:63)
	at io.helidon.security.Security.audit(Security.java:232)
	at io.helidon.security.SecurityContextImpl.audit(SecurityContextImpl.java:179)
	at io.helidon.security.AuthorizationClientImpl.lambda$submit$1(AuthorizationClientImpl.java:84)
	at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:986)
	at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:970)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
	... 3 more`

Received AuditEvent:
`event.messageFormat()` returned "Provider %s, Description %s, Request %s. Subject %s. %s: %s"
`event.params` returned 5 params

--- a/security/security/src/main/java/io/helidon/security/AuditEvent.java
+++ b/security/security/src/main/java/io/helidon/security/AuditEvent.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -84,7 +84,7 @@ public interface AuditEvent {
      * Gets the message format of this {@code AuditEvent} to be used with
      * {@link String#format(String, Object...)}.
      *
-     * @return English message format (this is a fallback if internationalization is not configured.
+     * @return English message format (this is a fallback if internationalization is not configured).
      */
     String messageFormat();
 

--- a/security/security/src/main/java/io/helidon/security/AuthorizationClientImpl.java
+++ b/security/security/src/main/java/io/helidon/security/AuthorizationClientImpl.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -86,6 +86,7 @@ final class AuthorizationClientImpl implements SecurityClient<AuthorizationRespo
                             "Provider %s, Description %s, Request %s. Subject %s. %s: %s")
                                           .addParam(AuditEvent.AuditParam
                                                             .plain("provider", providerInstance.getClass().getName()))
+                                          .addParam(AuditEvent.AuditParam.plain("description", "Audit failure"))
                                           .addParam(AuditEvent.AuditParam.plain("request", this))
                                           .addParam(AuditEvent.AuditParam.plain("subject",
                                                                                 context.user()

--- a/security/security/src/main/java/io/helidon/security/DefaultAuditProvider.java
+++ b/security/security/src/main/java/io/helidon/security/DefaultAuditProvider.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -105,8 +105,13 @@ final class DefaultAuditProvider implements AuditProvider {
                                                               throwable), () -> auditLogger.log(level, finalMsg));
     }
 
-    private String formatMessage(AuditEvent event) {
-        return String.format(event.messageFormat(), toObjectParams(event.params()));
+    String formatMessage(AuditEvent event) {
+        try {
+            return String.format(event.messageFormat(), toObjectParams(event.params()));
+        } catch (Exception e) {
+            // problem with the format
+            return "Formatting failed for format: " + event.messageFormat() + ", parameters: " + event.params();
+        }
     }
 
     private Object[] toObjectParams(List<AuditEvent.AuditParam> parameters) {
@@ -115,7 +120,7 @@ final class DefaultAuditProvider implements AuditProvider {
 
         for (AuditEvent.AuditParam param : parameters) {
             if (param.isSensitive()) {
-                result.add(param.name() + " (sensitive)");
+                result.add("(sensitive)");
             } else {
                 result.add(param.value().orElse("null"));
             }

--- a/security/security/src/test/java/io/helidon/security/AuditEventTest.java
+++ b/security/security/src/test/java/io/helidon/security/AuditEventTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -44,7 +44,7 @@ public class AuditEventTest {
     }
 
     @Test
-    public void testSensitiveParam() {
+    void testSensitiveParam() {
         String name = "paramName";
         String value = "sensitiveValue";
 

--- a/security/security/src/test/java/io/helidon/security/DefaultAuditProviderTest.java
+++ b/security/security/src/test/java/io/helidon/security/DefaultAuditProviderTest.java
@@ -1,0 +1,99 @@
+/*
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.security;
+
+import java.util.List;
+import java.util.Optional;
+
+import io.helidon.config.Config;
+
+import org.junit.jupiter.api.Test;
+
+import static io.helidon.common.CollectionsHelper.listOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.startsWith;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * Unit test for {@link DefaultAuditProvider}.
+ */
+class DefaultAuditProviderTest {
+    @Test
+    void testMessageFormatting() {
+        DefaultAuditProvider provider = DefaultAuditProvider.create(Config.empty());
+        AuditEvent event = createEvent("Unit tests: first: \"%s\"; second: \"%s\"",
+                                       listOf(AuditEvent.AuditParam.plain("first", "data"),
+                                              AuditEvent.AuditParam.sensitive("second", "secret")));
+
+        String message = provider.formatMessage(event);
+
+        assertThat(message, is("Unit tests: first: \"data\"; second: \"(sensitive)\""));
+    }
+
+    @Test
+    void testMessageFormattingTooManyParams() {
+        DefaultAuditProvider provider = DefaultAuditProvider.create(Config.empty());
+        AuditEvent event = createEvent("Unit tests: first: \"%s\"; second: \"%s\"",
+                                       listOf(AuditEvent.AuditParam.plain("first", "data"),
+                                              AuditEvent.AuditParam.sensitive("second", "secret"),
+                                              AuditEvent.AuditParam.plain("third", "thirdData")));
+
+        String message = provider.formatMessage(event);
+
+        assertThat(message, is("Unit tests: first: \"data\"; second: \"(sensitive)\""));
+    }
+
+    @Test
+    void testMessageFormattingNotEnoughParams() {
+        DefaultAuditProvider provider = DefaultAuditProvider.create(Config.empty());
+        AuditEvent event = createEvent("Unit tests: first: \"%s\"; second: \"%s\"",
+                                       listOf(AuditEvent.AuditParam.plain("first", "data")));
+
+        String message = provider.formatMessage(event);
+
+        assertThat(message, startsWith("Formatting failed for format: Unit tests: first: \"%s\"; second: \"%s\", parameters: "));
+    }
+
+    private AuditEvent createEvent(String messageFormat, List<AuditEvent.AuditParam> params) {
+        return new AuditEvent() {
+            @Override
+            public String eventType() {
+                return "unit-test";
+            }
+
+            @Override
+            public Optional<Throwable> throwable() {
+                return Optional.empty();
+            }
+
+            @Override
+            public List<AuditParam> params() {
+                return params;
+            }
+
+            @Override
+            public String messageFormat() {
+                return messageFormat;
+            }
+
+            @Override
+            public AuditSeverity severity() {
+                return null;
+            }
+        };
+    }
+}

